README.org
#+TITLE: GPTAI.el
#+AUTHOR: Anton Hibl

* an OpenAI API for Emacs
[[https://melpa.org/#/gptai][file:https://melpa.org/packages/gptai-badge.svg]]


This allows for easy communication between emacs and the openAI API
platform; allows using all of the available models and integrates cleanly with
emacs toolings. Here is the basic things you will need for set-up:

- Install the package from MELPA using:

  #+begin_src emacs-lisp
      (package-install 'gptai)
  #+end_src

  then put this in your init.el:

  #+begin_src emacs-lisp
    (require 'gptai)
    ;; set configurations
    (setq gptai-model "<MODEL-HERE>") 
    (setq gptai-username "<USERNAME-HERE>")
    (setq gptai-api-key "<API-KEY-HERE>")
    ;; set keybindings optionally
    (global-set-key (kbd "C-c o") 'gptai-send-query)
  #+end_src

To fill out the details of the configuration section

- Define the desired model to use; available models can be found by running
  ~gptai-list-models~ , it will also display this list in the *gptai*
  buffer. Use this to choose a model and set ~gptai-model~ (text-davinci-003 is
  a good default model).
- Define your OpenAI username.
- Define your API key (your own key can be obtained from [[https://platform.openai.com/account/api-keys][OpenAI API Keys]])

After that optionally fill out the init section with some keybindings.

** Extending the Package

Check out the [[https://github.com/antonhibl/gptai/wiki][wiki]] for examples on how to easily extend this package to your uses

** Text Querying

You can send textual queries to different models of openAI using the
functions:

- ~gptai-send-query~

  Prompts in the minibuffer a query and returns the output at the point.
  
- ~gptai-send-query-from-selection~

  Sends the text in the current selection to openAI as the prompt to the openAI
  API model you specified in your configurations, returns the output in place of
  the original selection.

- ~gptai-send-query-from-buffer~

  Sends the current buffer as the prompt to the openAI API model you specified
  in your configurations, returns the output in place of the original buffer.

- ~gptai-spellcheck-text-from-selection~

  Sends the selected text for spellchecking to openAI and replaces the selection
  with the corrected text.

- ~gptai-elaborate-on-text-from-selection~

  Sends the selected text for further elaboration by openAI and replaces the
  original selection with the improved text.

** Image Querying

You can generate images with the DALL-E engine using:

- ~gptai-send-image-query~

  Prompts the user in the minibuffer for

  - A prompt for the command
  - How many images you want to generate
  - What size they should be
  - Where you want to store them on the disk

  It will then generate those images at the API endpoint, then use curl to
  download those images to your specified directory path. Once it is done, if
  one image was downloaded it will open it in a new buffer for viewing,
  otherwise if more were downloaded it will simply display success when done.

** Code Querying and Generation

Using the ~gptai-code-query~ and ~gptai-code-query-from-selection~ functions,
you are able to generate code from instructive prompts in specified languages,
both of these functions by default will prompt the user for a language and
return the returned code in-place at your selection or point in the buffer.

- ~gptai-code-query~

  Prompts the user for instructions and language to use; then returns the output
  code that was generated at the point.

- ~gptai-code-query-from-selection~

  Uses the selection text as the instructions and prompts for a language to use;
  then returns the code that was generated in place of the original selection.

- ~gptai-explain-code-from-selection~

  Explain the code from the selection, return output above selection.

- ~gptai-fix-code-from-selection~

  Fix and debug the code from the selection, return the output in place of the
  original selection.

- ~gptai-document-code-from-selection~

  Document and describe the code from the selection, return output above
  selection.

- ~gptai-optimize-code-from-selection~

  Optimizes and refactors code from selection, returns output in place of the
  original selection.

- ~gptai-improve-code-from-selection~

  Improves and extends on code from selection, returns output in place of
  original selection.

** Using the 3.5-turbo models

The newer models of gpt-3.5 and gpt-4 requires a different query structure, this
means that to use these models there needed to be a seperate query handler; this
is achieved with ~gpt-turbo.el~. Using the turbo model is as easy as calling
~gptai-turbo-query~ while passing it a text query input either interactively or
pragrammatically. Here is an example:

#+begin_src emacs-lisp
  (gptai-turbo-query "This is a prompt")
#+end_src

this functions basically the same as gptai-send-query, it just limits you to the
newer 3.5 turbo, and 4 models for GPT language models instead of giving more
ability to choose what models you are using witht hes tandard query handler.

** Contributing

Feel free to make a PR with improvements, all PRs should include your changes as
well as a addition to the CHANGELOG.md file noting any important changes for
users to be aware of.


CHANGELOG.md
# Changelog

All changes that impact users of this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!---
This document is intended for the users of astrolisp. Changes to things
like tests should not be noted in this document.
When updating this file for a PR, add an entry for your change under Unreleased
and one of the following headings:

 - Added - for new features.
 - Changed - for changes in existing functionality.
 - Deprecated - for soon-to-be removed features.
 - Removed - for now removed features.
 - Fixed - for any bug fixes.
 - Security - in case of vulnerabilities.
-->

## Changes

### Added
**Date:** *Feb 24th, 2023*
**Author:** *Anton Hibl*

Added a changelog to help users keep track of changes more clearly.

### Changed
**Date:** *Feb 26th, 2023*
**Author:** *Anton Hibl*

Changed gptai-model and gptai-api-key so that they are able to be set through
the customization menu

### Added 
**Date:** *Mar 17th, 2023*
**Author:** *Anton Hibl*


Added gptai-turbo which includes basic query functionality for the gpt-3.5-turbo
model; planned to be extended into a larger interface and chat interface
soon. Tests have been included in the tests directory for development purposes,
but they are not installed during package-install as they are intended solely
for development and not for end users. Including them would unnecessarily
increase the package size.


gptai-test.el
;;; gptai-test.el --- Test Suite for GPTAI -*- lexical-binding: t; -*-

;; Copyright (C) 2023 Hibl, Anton

;; Author: Anton Hibl <antonhibl11@gmail.com>
;; URL: https://github.com/antonhibl/gptai
;; Keywords: comm, convenience
;; Version: 1.0.4
;; Package-Requires: ((emacs "24.1"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; A suite of tests for the gptai functions to ensure they work as
;; expected.  To run these tests either:

;; A. M-x ert will open the ERT interactive test runner, where you can select
;; the tests you want to run.  Press t to run all the tests, or use r to run a
;; specific test by its name.  The test results will be displayed in the ERT test
;; runner buffer.

;; B. By running (ert-run-tests-batch-and-exit), the tests will be executed in
;; batch mode, and the test results will be displayed in the *Messages*
;; buffer.  If all tests pass, you will see the message All tests passed.

;;; Code:

(require 'ert)
(require 'gptai)

;;;###autoload
(defvar gptai-test-mock-api-response
  '((choices . [((text . "Mock API response text.") (index . 0))]))
  "A mock API response for testing purposes.")

;;;###autoload
(defun gptai-test-request-mock ()
  "Simulates a GPTAI request without making an actual API call.
Argument GPTAI-PROMPT is the prompt to send to the API."
  gptai-test-mock-api-response)

;;;###autoload
(defun gptai-test-send-query-mock ()
  "Sends a query to the mock API and insert the response at the point.
Argument GPTAI-PROMPT prompt."
  (let ((response (gptai-test-request-mock)))
    (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
      (if text
          (insert text)
        (error
         "Response doesn't contain text data")))))

;;;###autoload
(ert-deftest gptai-test-request-mock ()
  (should (equal (gptai-test-request-mock) gptai-test-mock-api-response)))

;;;###autoload
(ert-deftest gptai-test-send-query-mock ()
  (with-temp-buffer
    (gptai-test-send-query-mock)
    (should (equal (buffer-string) "Mock API response text."))))

(provide 'gptai-test)
;;; gptai-test.el ends here



gptai-tests.el
;;; gptai-test.el --- test suite for gptai -*- lexical-binding: t; -*-

;; Copyright (C) 2023 Hibl, Anton

;; Author: Anton Hibl <antonhibl11@gmail.com>
;; URL: https://github.com/antonhibl/gptai
;; Keywords: comm, convenience
;; Version: 0.1.0
;; Package-Requires: ((emacs "24.1"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; A suite of tests for the gptai functions to ensure they work as
;; expected. To run these tests either:

;; A. M-x ert will open the ERT interactive test runner, where you can select
;; the tests you want to run. Press t to run all the tests, or use r to run a
;; specific test by its name. The test results will be displayed in the ERT test
;; runner buffer. 

;; B. By running (ert-run-tests-batch-and-exit), the tests will be executed in
;; batch mode, and the test results will be displayed in the *Messages*
;; buffer. If all tests pass, you will see the message All tests passed.

;;; Code:

(require 'ert)
(require 'gptai) ; assuming the provided code is saved in gptai.el

(defun gptai-mock-request (gptai-prompt)
  "Mock GPTAI-request function to simulate API call without actually calling it.
Argument GPTAI-PROMPT is the prompt to send to the API."
  (let ((response `((id . "test-id")
                    (object . "test-object")
                    (created . 1234567890)
                    (model . ,gptai-model)
                    (usage . ((prompt_tokens . 10)
                              (completion_tokens . 20)
                              (total_tokens . 30)))
                    (choices . [((text . "Test response for prompt: ")
                                 (index . 0)
                                 (finish_reason . "stop")
                                 (logprobs . nil))]))))
    response))

(defun gptai-test-send-query (gptai-prompt)
  "Test function for gptai-send-query using mock API response.
Argument GPTAI-PROMPT prompt."
  (let ((response (gptai-mock-request gptai-prompt)))
    (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
      (if text
          (insert text)
        (error "Response doesn't contain text data")))))

(ert-deftest gptai-test-query-functions ()
  "Test gptai-send-query and other query functions using the mock request."
  (let ((gptai-prompt "Test prompt")
        (gptai-request-function #'gptai-mock-request))

    ;; Test gptai-send-query
    (with-temp-buffer
      (gptai-test-send-query gptai-prompt)
      (should (equal (buffer-string) "Test response for prompt: ")))

    ;; Test gptai-send-query-region
    (with-temp-buffer
      (insert gptai-prompt)
      (goto-char (point-min))
      (push-mark)
      (goto-char (point-max))
      (activate-mark)
      (gptai-send-query-region)
      (should (equal (buffer-string) "Test response for prompt: ")))

    ;; Test gptai-send-chat
    (with-temp-buffer
      (gptai-send-chat gptai-prompt)
      (should (equal (buffer-string) "Test response for prompt: ")))

    ;; Test gptai-open-chat
    (gptai-open-chat gptai-prompt)
    (should (equal (buffer-string) (concat gptai-prompt "\n\nChatGPT:Test response for prompt: ")))

    ;; Test gptai-continue-chat
    (with-current-buffer "*chat-gptai*"
      (gptai-continue-chat)
      (should (equal (buffer-string) (concat gptai-prompt "\n\nChatGPT:Test response for prompt: "
                                             "Test response for prompt: ")))

    ;; Test gptai-send-query-buffer
    (with-temp-buffer
      (insert gptai-prompt)
      (gptai-send-query-buffer)
      (should (equal (buffer-string) "Test response for prompt: "))))))

(provide 'gptai-test)
;;; gptai-test.el ends here


gptai-turbo-test.el
;;; gptai-turbo-test.el --- Test Suite for gptai-turbo.el -*- lexical-binding: t; -*-

;; Copyright (C) 2023 Hibl, Anton

;; Author: Anton Hibl <antonhibl11@gmail.com>
;; URL: https://github.com/antonhibl/gptai
;; Keywords: comm, convenience
;; Version: 1.0.4
;; Package-Requires: ((emacs "24.1"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; A suite of tests for the gptai-turbo functions to ensure they work as
;; expected.  To run these tests either:

;; A. M-x ert will open the ERT interactive test runner, where you can select
;; the tests you want to run.  Press t to run all the tests, or use r to run a
;; specific test by its name.  The test results will be displayed in the ERT test
;; runner buffer.

;; B. By running (ert-run-tests-batch-and-exit), the tests will be executed in
;; batch mode, and the test results will be displayed in the *Messages*
;; buffer.  If all tests pass, you will see the message All tests passed.

;; These tests use a mock API response and a test function that simulates
;; gptai-turbo-request without making an actual API call to ensure that the
;; tests do not depend on the availability of the API and to avoid rate-limiting
;; issues.

;;; Code:

(require 'ert)
(require 'gptai)
(require 'gptai-turbo)

;; Mock API response
(defvar gptai-turbo-test-mock-api-response
  '((id . "chatcmpl-6uaMfmuqdi2R0Zj5HSgKvcUjQceeJ")
    (object . "chat.completion")
    (created . 1678944173)
    (model . "gpt-3.5-turbo-0301")
    (usage (prompt_tokens . 11) (completion_tokens . 41) (total_tokens . 52))
    (choices . [((message (role . "assistant") (content . "This is a test response.")) (finish_reason . "stop") (index . 0))])))

;;;###autoload
(defun gptai-turbo-test-request (prompt)
  "Simulates `gptai-turbo-request' without making an actual API call.
Argument PROMPT is the prompt to send to the API, but it's not used in this mock function."
  (let ((first-choice (elt (cdr (assoc 'choices gptai-turbo-test-mock-api-response)) 0)))
    (cdr (assoc 'content (cdr (assoc 'message first-choice))))))

;;;###autoload
(ert-deftest gptai-turbo-test-request-test ()
  "Test `gptai-turbo-request' with a mock response."
  (let* ((prompt "Test prompt")
         (gptai-turbo-request (symbol-function 'gptai-turbo-request)))
    (fset 'gptai-turbo-request (symbol-function 'gptai-turbo-test-request))
    (should (string= (gptai-turbo-request prompt) "This is a test response."))
    (fset 'gptai-turbo-request gptai-turbo-request)))

;;;###autoload
(ert-deftest gptai-turbo-test-response-test ()
  "Test `gptai-turbo-response' with a mock response."
  (let* ((prompt "Test prompt")
         (gptai-turbo-request (symbol-function 'gptai-turbo-request))
         (buffer (generate-new-buffer "*gptai-turbo-response-test*")))
    (fset 'gptai-turbo-request (symbol-function 'gptai-turbo-test-request))
    (with-current-buffer buffer
      (gptai-turbo-response prompt)
      (should (string= (buffer-string) "This is a test response.")))
    (fset 'gptai-turbo-request gptai-turbo-request)
    (kill-buffer buffer)))

(provide 'gptai-turbo-test)
;;; gptai-turbo-test.el ends here


turbo.el
(defun gptai-turbo-request (gptai-prompt)
  "Sends a request to OpenAI API's gpt-3.5-turbo endpoint and returns the response.
Argument GPTAI-PROMPT is the prompt to send to the API."
  (when (null gptai-api-key)
    (error "OpenAI API key is not set"))

  (let* ((url-request-method "POST")
         (url-request-extra-headers
          `(("Content-Type" . "application/json")
            ("Authorization" . ,(format "Bearer %s" gptai-api-key))))
         (url-request-data
          (json-encode `(("model" . "gpt-3.5-turbo")
                         ("messages" . [((role . "user") (content . ,gptai-prompt))])
                         ("temperature" . 0.7))))
         (url "https://api.openai.com/v1/chat/completions")
         (buffer (url-retrieve-synchronously url nil 'silent))
         response)

    (message "Sending request to OpenAI API using model 'gpt-3.5-turbo'")

    (if buffer
      (with-current-buffer buffer
        (goto-char url-http-end-of-headers)
        (condition-case gptai-err
            (progn
              (setq response (json-read))
              (if (assoc 'error response)
                  (error (cdr (assoc 'message (cdr (assoc 'error response)))))
                (let ((first-choice (elt (cdr (assoc 'choices response)) 0))) ; Extract the first choice
                  (cdr (assoc 'content (cdr (assoc 'message first-choice))))))) ; Get the 'content' field of the first choice
          (error (error "Error while parsing OpenAI API response: %s"
                        (error-message-string gptai-err)))))
      (error "Failed to send request to OpenAI API"))))

(defun gptai-turbo-response (gptai-prompt)
  "Sends a request to gpt-3.5-turbo and inserts response at the current point.
Argument GPTAI-PROMPT prompt to be sent."
  (interactive "sEnter your prompt: ")
  (let ((response (gptai-turbo-request gptai-prompt)))
    (insert response)))


gptai.el
;;; gptai.el --- Integrate with the OpenAI API -*- lexical-binding: t; -*-

;; Copyright (C) 2023 Hibl, Anton

;; Author: Anton Hibl <antonhibl11@gmail.com>
;; URL: https://github.com/antonhibl/gptai
;; Keywords: comm, convenience
;; Version: 1.0.5
;; Package-Requires: ((emacs "24.1"))

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;
;;     ____ ____ _____  _    ___       _
;;    / ___|  _ \_   _|/ \  |_ _|  ___| |
;;   | |  _| |_) || | / _ \  | |  / _ \ |
;;   | |_| |  __/ | |/ ___ \ | | |  __/ |
;;    \____|_|    |_/_/   \_\___(_)___|_|
;;  
;; This is intended to allow for development and programming queries into the
;; OpenAI API.  This allows for sending queries straight from Emacs directly into
;; various models of OpenAI's platform.  It is a barebones implementation of a
;;     wrapper around the API focused on achieving extensibility.

;; Configurations that are required are listed as follows:
;;
;; - Define the desired model to use (available models can be found by running
;;   gptai-list-models which will populate the gptai-models variable with the
;;   list of all available models, it will also display this list in the gptai
;;   buffer).
;;
;; - Define your OpenAI API key.
;;
;; - Optionally define keybindings for sending various queries easily.
;;
;; An example of these configurations after installing from MELPA is shown
;; below:
;;
;;   (require 'gptai)
;;   ;; set standard configurations
;;   (setq gptai-model "<MODEL-HERE>")
;;   (setq gptai-username "<USERNAME-HERE>")
;;   (setq gptai-api-key "<API-KEY-HERE>")
;;   ;; set keybindings optionally
;;   (global-set-key (kbd "C-c o") 'gptai-send-query)
;;

;;; Code:

;;; Customization
(defgroup gptai nil
  "Use the openAI API."
  :prefix "gptai-"
  :group 'comm
  :link '(url-link :tag "Repository" "https://github.com/antonhibl/gptai"))

;; dependencies
(require 'url)
(require 'json)

;; default values for local variables
(defvar gptai-base-url "https://api.openai.com/v1/completions")
(defvar gptai-chat-url "https://api.openai.com/v1/chat/completions")
(defcustom gptai-model ""
  "API Model for OpenAI."
  :type 'string
  :group 'gptai)
(defcustom gptai-api-key ""
  "API key for OpenAI."
  :type 'string
  :group 'gptai)
(defvar url-http-end-of-headers)
(defvar gptai-image)
(defvar gptai-images)
(defvar gptai-indn)
(defvar gptai-index)

(defun gptai-request (gptai-prompt)
  "Sends a request to OpenAI API and return the response.
Argument GPTAI-PROMPT is the prompt to send to the API."
  (when (null gptai-api-key)
    (error "OpenAI API key is not set"))

  (let* ((url-request-method "POST")
         (url-request-extra-headers
          `(("Content-Type" . "application/json")
            ("Authorization" . ,(format "Bearer %s" gptai-api-key))))
         (url-request-data
          (json-encode `(("model" . ,gptai-model)
                         ("prompt" . ,gptai-prompt)
                         ("temperature" . 0.7)
                         ("max_tokens" . 1000))))
         (buffer (url-retrieve-synchronously gptai-base-url nil 'silent))
         response)

    (message "Sending request to OpenAI API using model '%s'" gptai-model)

    (if buffer
        (with-current-buffer buffer
          (goto-char url-http-end-of-headers)
          (condition-case gptai-err
              (progn
                (setq response (json-read))
                (if (assoc 'error response)
                    (error (cdr (assoc 'message (cdr (assoc 'error response)))))
                  response))
            (error (error "Error while parsing OpenAI API response: %s"
                          (error-message-string gptai-err)))))
      (error "Failed to send request to OpenAI API"))))

(defun gptai-send-query (gptai-prompt)
  "Sends a query to OpenAI API and insert the response text at the current point.
Argument GPTAI-PROMPT prompt."
  (interactive
   (list (read-string "Prompt: ")))
  (let ((response (gptai-request gptai-prompt)))
    (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
      (if text
          (insert text)
        (error
         "Response doesn't contain text data")))))

(defun gptai-send-query-region ()
  "Sends query to OpenAI API using region, replace region w/ response."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                          (buffer-substring-no-properties
                           (region-beginning)
                           (region-end))
                        (read-string "Query: "))))
    (let ((response (gptai-request gptai-prompt)))
      (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
        (delete-region (region-beginning)
                       (region-end))
        (insert text)))))

(defun gptai-send-query-buffer (&optional buffer-name)
  "Sends a query to OpenAI API using the buffer as a prompt.
Optional argument BUFFER-NAME buffer to send."
  (interactive
   (list (read-buffer "Buffer: " (current-buffer))))
  (let ((gptai-prompt (with-current-buffer buffer-name
                        (buffer-substring
                         (point-min)
                         (point-max)))))
    (let ((response (gptai-request gptai-prompt)))
      (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response))
                                         0)))))
        (with-current-buffer buffer-name
          (erase-buffer)
          (insert text)
          (display-buffer (current-buffer) t))))))

(defun gptai-spellcheck-region ()
  "Sends query to OpenAI API to spellcheck the region."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Text: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "Spellcheck this text: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text))))))

(defun gptai-elaborate-on-region ()
  "Sends query to OpenAI API to elaborate on the region."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Text: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "Elaborate on this text: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text))))))

(defun gptai-code-query-region (gptai-language)
  "Sends instructions to OpenAI API to code in a language.
Argument GPTAI-INSTRUCTIONS code instructions for query.
Argument GPTAI-LANGUAGE language to generate."
  (interactive
   (list (read-string "Language: ")))
  (let ((gptai-code-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Code: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "%s(%s)" gptai-code-prompt gptai-language))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text))))))

(defun gptai-code-query (gptai-instructions gptai-language)
  "Sends instructions to OpenAI API to code in a language.
Argument GPTAI-INSTRUCTIONS code instructions for query.
Argument GPTAI-LANGUAGE language to generate."
  (interactive
   (list (read-string "Instructions: ")
         (read-string "Language: ")))
  ;; place code in current buffer
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "%s(%s)" gptai-instructions gptai-language))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text)))))

(defun gptai-explain-code-region ()
  "Sends region to OpenAI API to explain."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Code: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "Explain the following: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (insert text)
          (insert "\n\n"))))))

(defun gptai-document-code-region ()
  "Sends region to OpenAI API to write documentation."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Code: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "Please write the documentation for the following code: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (insert text)
          (insert "\n\n"))))))

(defun gptai-optimize-code-region ()
  "Sends region to OpenAI API to optimize."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Code: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "Optimize and refactor the following code: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text))))))


(defun gptai-improve-code-region ()
  "Sends region to OpenAI API to improve."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Code: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "Improve and extend the following code: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text))))))

(defun gptai-fix-code-region ()
  "Sends region to OpenAI API to fix a bug."
  (interactive)
  (let ((gptai-prompt (if (use-region-p)
                    (buffer-substring-no-properties (region-beginning) (region-end))
                  (read-string "Code: "))))
    (with-current-buffer (current-buffer)
      (let ((response (gptai-request (format "There is a bug in the following function, please help me fix it: %s" gptai-prompt))))
        (let ((text (cdr (assoc 'text (elt (cdr (assoc 'choices response)) 0)))))
          (delete-region (region-beginning) (region-end))
          (insert text))))))

(defun gptai-send-image-query (prompt n size filepath)
  "Sends a query to the OpenAI Image Generation API.
Argument PROMPT prompt to send.
Argument N number of images to generate.
Argument SIZE size of images.
Argument FILEPATH filepath to download to."
  (interactive
   (list (read-string "Prompt: ")
         (read-number "Number of images to generate: " 1)
         (read-string "Image size (e.g. 1024x1024): " "1024x1024")
         (read-directory-name "Enter output directory: " "~/Pictures")))
  (when (null gptai-api-key)
    (error "OpenAI API key is not set"))
  (let* ((url "https://api.openai.com/v1/images/generations")
         (url-request-method "POST")
         (url-request-extra-headers
          `(("Content-Type" . "application/json")
            ("Authorization" . ,(format "Bearer %s" gptai-api-key))))
         (url-request-data
          (json-encode `(("prompt" . ,prompt)
                         ("n" . ,n)
                         ("size" . ,size)))))
    (message "Sending request to OpenAI Image Generation API")
    (condition-case err
        (with-current-buffer
            (url-retrieve-synchronously url nil 'silent)
          (goto-char url-http-end-of-headers)
          (let ((response (json-read)))
            (when (assoc 'error response)
              (error (cdr (assoc 'message (cdr (assoc 'error response))))))
            (with-current-buffer (get-buffer-create "*openai*")
              (erase-buffer)
              (insert (format "Generated image URLs:\n"))
              (setq gptai-indn n)
              (setq gptai-image (let ((gptai-indn (length (cdr (assoc 'data response))))
                    (urls '()))
                (dotimes (x gptai-indn)
                  (push (cdr (assoc 'url (elt (cdr (assoc 'data response)) x))) urls))
                (reverse urls)))
              (insert (format "%s\n"
                              (let ((gptai-indn (length (cdr (assoc 'data response))))
                                    (urls '()))
                                (dotimes (x gptai-indn)
                                  (push (cdr (assoc 'url (elt (cdr (assoc 'data response)) x))) urls))
                                (reverse urls)))))
            (switch-to-buffer-other-window (current-buffer)))
          (setq gptai-index 0)
          (let ((gptai-images gptai-image))
              (dolist (gptai-image gptai-images)
                (async-shell-command (format "curl '%s' > %s/%s_%d.png" gptai-image filepath (format-time-string "%T") gptai-index))
                (sleep-for 2 500)
                (setq gptai-index (+ gptai-index 1))))
            (message "Finished downloading %d images to %s" n filepath))
      (error (error "Error while sending request to OpenAI Image Generation API: %s"
                    (error-message-string err))))))

(defun gptai-list-models ()
  "Retrieves a list of currently available GPT-3 models from OpenAI."
  (interactive)
  (with-current-buffer (get-buffer-create "*gptai-models*")
    (erase-buffer)
    (async-shell-command (format "curl https://api.openai.com/v1/models \
    -H 'Authorization: Bearer %s'" gptai-api-key) "*gptai-models*" "*Messages*")
    (goto-char (point-min))
    (re-search-forward "^.*object.*$")
    (delete-region (point-min) (point))
    (pop-to-buffer (current-buffer))))

(provide 'gptai)
;;; gptai.el ends here
